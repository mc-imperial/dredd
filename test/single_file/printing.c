#include <stdio.h>

int main() {
  printf("%s\n", "A\n");
  printf("%s\n", "B\n");
  printf("%s\n", "C\n");
  printf("%s\n", "D\n");
  printf("%s\n", "E\n");
  printf("%s\n", "F\n");
  printf("%s\n", "G\n");
}
