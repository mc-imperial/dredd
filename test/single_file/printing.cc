#include <iostream>

int main() {
  std::cout << "A\n";
  std::cout << "B\n";
  std::cout << "C\n";
  std::cout << "D\n";
  std::cout << "E\n";
  std::cout << "F\n";
  std::cout << "G\n";
}
