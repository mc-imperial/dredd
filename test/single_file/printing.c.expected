#include <stdio.h>

#include <inttypes.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#ifdef _MSC_VER
#define thread_local __declspec(thread)
#elif __APPLE__
#define thread_local __thread
#else
#include <threads.h>
#endif

static thread_local int __dredd_some_mutation_enabled = 1;
static bool __dredd_enabled_mutation(int local_mutation_id) {
  static thread_local int initialized = 0;
  static thread_local uint64_t enabled_bitset[1];
  if (!initialized) {
    int some_mutation_enabled = 0;
    const char* dredd_environment_variable = getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable) {
      char* temp = malloc(strlen(dredd_environment_variable) + 1);
      strcpy(temp, dredd_environment_variable);
      char* token;
      token = strtok(temp, ",");
      while(token) {
        int value = atoi(token);
        int local_value = value - 0;
        if (local_value >= 0 && local_value < 7) {
          enabled_bitset[local_value / 64] |= ((uint64_t) 1 << (local_value % 64));
          some_mutation_enabled = 1;
        }
        token = strtok(NULL, ",");
      }
      free(temp);
    }
    initialized = 1;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return enabled_bitset[local_mutation_id / 64] & ((uint64_t) 1 << (local_mutation_id % 64));
}

int main() {
  if (!__dredd_enabled_mutation(0)) { printf("%s\n", "A\n"); }
  if (!__dredd_enabled_mutation(1)) { printf("%s\n", "B\n"); }
  if (!__dredd_enabled_mutation(2)) { printf("%s\n", "C\n"); }
  if (!__dredd_enabled_mutation(3)) { printf("%s\n", "D\n"); }
  if (!__dredd_enabled_mutation(4)) { printf("%s\n", "E\n"); }
  if (!__dredd_enabled_mutation(5)) { printf("%s\n", "F\n"); }
  if (!__dredd_enabled_mutation(6)) { printf("%s\n", "G\n"); }
}
